feat: Focus first match on search

MJRuskin opened the pull request:
## What does it do?

When searching, if the current focus is not a match, it will focus the first match.

## Fixes # (issue)

No issueNumber

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] Updated documentation (if applicable)
- [ ] Added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] My changes generate no new warnings

MJRuskin commented:
Just realised this PR changes how some of the navigation tests would work - since the first match is automatically focused before navigating with the arrow keys.

For fixing the tests, would you prefer me to: 
- "correct" the original tests and create an additional test to demonstrate the new feature
or
- integrate a check for this new feature into the original two tests and correct the outcome

(I am assuming the second option is more preferable)

mrchief commented:
Hey! Thanks, for sending this.

> When searching, if the current focus is not a match, it will focus the first match.

So the reason we do not focus anything right now is to let the user keep on searching. It'd be annoying if I'm typing something and I pause for a sec and then focus is lost from the input. 

I'm not sure automatically switching focus is a good idea in that case but I'm happy to hear what your thoughts are.

MJRuskin commented:
The search input will never lose focus. This change affects which item in the tree is highlighted.

My intention for the change was to allow users to type and then simply press the Enter key to quickly select the first match.

In the current state this requires the additional press of an arrow key to snap to the first match.

I can gladly make this an optional feature instead. I just felt the current implementation was inconsistent when it came to navigation - especially when going back and forth between searching and the normal view.

mrchief commented:
OK, I see. Yeah, it seems like this will be a good addition. I need to play with it a bit to see it in action but would love to hear what @ellinge thinks as well.

mrchief commented:
Can we also add some unit tests for this?

MJRuskin commented:
Just added some tests and fixed the ones that were affected by this change.

I have never used enzyme before, so please let me know if something needs fixing.

marceleq27 commented:
Hi, any updates in this PR? I was searching in docs for this feature and I came here :)

mrchief commented:
@marceleq27 haven't had time to dig into this. Will try to get it this weekend.
